### PR TITLE
testing: add handling of unknown providers and empty terraform

### PR DESCRIPTION
**Pull request by xlr-8**
This PR adds the handling of unknown provider (eg. vmware) and also cases of empty terraform plan/module. That allows to give better output when the cost estimated is '0.0'. Now instead of having '0.0' an error will be returned.

In order to improve the testing, some pricing dump got also injected as part of it.

Related to: https://github.com/cycloidio/terracost/issues/21

### CODE REVIEW

File: terraform/error.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package terraform
+
+import "fmt"
+
+// Errors that might be returned from procesing the HCL
+var (
+	ErrNoQueries       = fmt.Errorf("no terraform entities found, looks empty")
```
Comment (xescugc) [inline terraform/error.go:7]: Means empty or that we do not support any of the resources on the Plan/HCL?

**Review by xescugc — CHANGES_REQUESTED** [review]
I feel like the HCL case is not handled and that this is maybe not the best way to handle the case that we do not have 0.0.

There are 2 ways to do it:
* On the functions of the `cost.Plan`
* Initialization of the Plan if the `prior` and `planned` are totally empty.

Or following what you did you could also add the `if` onto the `terraform.ExtractQueriesFromHCL`

File: terraform/error.go (hunk 1 above)
Comment (xlr-8) [reply terraform/error.go:7]: Means no queries were found - eg something generates a diff where there is nothing to process or contains no files.

**Comment by xlr-8** [comment]
> I feel like the HCL case is not handled and that this is maybe not the best way to handle the case that we do not have 0.0.

Why? Both cases are tested on the PR though?

> There are 2 ways to do it:
>     * On the functions of the `cost.Plan`
>     * Initialization of the Plan if the `prior` and `planned` are totally empty.

HCL doesn't contain prior and plan costs. Plus as we are speaking of plan, you could have some prior queries, but no new ones.
 
> Or following what you did you could also add the `if` onto the `terraform.ExtractQueriesFromHCL`

I'm not sure to follow your point there.

**Comment by xlr-8** [comment]
I'm currently having issues with the pipeline to properly test it, so I'll remove the `need review` label.

**Comment by xescugc** [comment]
> Why? Both cases are tested on the PR though?

IDK why I did only see the `NoQuereis` on the State and not the HCL case

> HCL doesn't contain prior and plan costs. Plus as we are speaking of plan, you could have some prior queries, but no new ones.

What I don't like about your solution is that it's the same logic in 2 places, and the `Queries` itself are of no "use" until you put them into a `State`. And both of them use the `State` to calculate so I would move that logic to the State initialization, so then it's not HCL nor Plan but common place, and also it's where the actual calculations are made so it would make sense to be there.

Also as the query returns s `[]query` you could calculate them from different files and then aggregate all the queries into 1 State which would validate if it's empty or not.

**Comment by xlr-8** [comment]
Done, pipeline has been fixed too to work with the dump.

**Review by xescugc — COMMENTED** [review]
rs

**Comment by xlr-8** [comment]
Done

File: estimation.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ func EstimateTerraformPlan(ctx context.Context, backend Backend, plan io.Reader,
 	}
 
 	priorQueries, err := tfplan.ExtractPriorQueries()
-	if err != nil {
+	if err != nil && err != terraform.ErrNoQueries {
```
Comment (xescugc) [inline estimation.go:27]: This is not needed no?

Comment (xlr-8) [reply estimation.go:27]: No, AFAIK this is needed in case there are no prior queries.

**Comment by xlr-8** [comment]
I've updated the Makefile/docker-compose file to ease the testing locally/in the CI

File: estimation.go (hunk 2 above)
Comment (xescugc) [reply estimation.go:27]: But the `ExtractPriorQueries` doeds not return this error anymore, this was useful before but now it's not.

Comment (xlr-8) [reply estimation.go]: Indeed, done

**Review by xescugc — APPROVED** [review]
RS + merge

**Pull request merged**